core/corehttp/gateway_handler.go

```diff
@@ -273,116 +275,84 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
+	if rsegs[0] == ipnsPathPrefix {
+		webError(w, "putHandler: updating named entries not supported", errors.New("WritableGateway: ipns put not supported"), http.StatusBadRequest)
 		return
 	}
 
 	var newnode *dag.Node
-	if pathext[len(pathext)-1] == '/' {
+	if rsegs[len(rsegs)-1] == "QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn" {
 		newnode = uio.NewEmptyDirectory()
 	} else {
-		newnode, err = i.newDagFromReader(r.Body)
+		putNode, err := i.newDagFromReader(r.Body)
 		if err != nil {
-			webError(w, "Could not create DAG from request", err, http.StatusInternalServerError)
+			webError(w, "putHandler: Could not create DAG from request", err, http.StatusInternalServerError)
 			return
 		}
+		newnode = putNode
 	}
 
-	ctx, cancel := context.WithCancel(i.node.Context())
-	defer cancel()
-
-	ipfsNode, err := core.Resolve(ctx, i.node, path.Path(urlPath))
-	if err != nil {
-		// FIXME HTTP error code
-		webError(w, "Could not resolve name", err, http.StatusInternalServerError)
-		return
+	var newPath string
+	if len(rsegs) > 1 {
+		newPath = filepath.Join(rsegs[2:]...)
 	}
 
-	k, err := ipfsNode.Key()
-	if err != nil {
-		webError(w, "Could not get key from resolved node", err, http.StatusInternalServerError)
-		return
-	}
-
-	h, components, err := path.SplitAbsPath(path.FromKey(k))
-	if err != nil {
-		webError(w, "Could not split path", err, http.StatusInternalServerError)
-		return
-	}
-
-	if len(components) < 1 {
-		err = fmt.Errorf("Cannot override existing object")
-		webError(w, "http gateway", err, http.StatusBadRequest)
-		return
-	}
+	var newkey key.Key
+	rnode, err := core.Resolve(i.node.Context(), i.node, rootPath)
+	switch ev := err.(type) {
+	case path.ErrNoLink:
+		// ev.Node < node where resolve failed
+		// ev.Name < new link
+		// but we need to patch from the root
+		rnode, err := i.node.DAG.Get(i.node.Context(), key.B58KeyDecode(rsegs[1]))
+		if err != nil {
+			webError(w, "putHandler: Could not create DAG from request", err, http.StatusInternalServerError)
+			return
+		}
 
-	tctx, cancel := context.WithTimeout(ctx, time.Minute)
-	defer cancel()
-	// TODO(cryptix): could this be core.Resolve() too?
-	rootnd, err := i.node.Resolver.DAG.Get(tctx, key.Key(h))
-	if err != nil {
-		webError(w, "Could not resolve root object", err, http.StatusBadRequest)
-		return
-	}
+		newPath := filepath.Join(rsegs[2:]...)
 
-	// resolving path components into merkledag nodes. if a component does not
-	// resolve, create empty directories (which will be linked and populated below.)
-	pathNodes, err := i.node.Resolver.ResolveLinks(tctx, rootnd, components[:len(components)-1])
-	if _, ok := err.(path.ErrNoLink); ok {
-		// Create empty directories, links will be made further down the code
-		for len(pathNodes) < len(components) {
-			pathNodes = append(pathNodes, uio.NewDirectory(i.node.DAG).GetNode())
+		e := dagutils.NewDagEditor(i.node.DAG, rnode)
+		err = e.InsertNodeAtPath(i.node.Context(), newPath, newnode, uio.NewEmptyDirectory)
```

Comment:
Awesome! should I do it in this PR? feels like.. ![](https://ipfs.io/ipfs/Qma21RdimUoYEK5JvyLfk6ZZJAhcDhkTbDSQcPnhQVhFcJ)